src/EVM/Fuzz.hs

```diff
@@ -0,0 +1,238 @@
+tryCexFuzz :: [Prop] -> Integer -> (Maybe (SMTCex))
+tryCexFuzz ps tries = unGen (testVals tries) (mkQCGen 0) 1337
+  where
+    vars = extractVars ps
+    bufs = extractBufs ps
+    stores = extractStorage ps
+    testVals :: Integer -> Gen (Maybe SMTCex)
+    testVals 0 = pure Nothing
+    testVals todo = do
+      varVals <- getVals vars
+      bufVals <- getBufs bufs
+      storeVals <- getStores stores
+      let
+        ret =  filterCorrectKeccak $ map (substituteEWord varVals . substituteBuf bufVals . substituteStores storeVals) ps
+        retSimp =  Expr.simplifyProps ret
+      if null retSimp then pure $ Just (SMTCex {
+                                    vars = varVals
+                                    , addrs = mempty
+                                    , buffers = bufVals
+                                    , store = storeVals
+                                    , blockContext = mempty
+                                    , txContext = mempty
+                                    })
+                                    else testVals (todo-1)
+
+-- Filter out PEq (Lit X) (keccak (ConcreteBuf Y)) if it's correct
+filterCorrectKeccak :: [Prop] -> [Prop]
+filterCorrectKeccak ps = filter checkKecc ps
+  where
+    checkKecc (PEq (Lit x) (Keccak (ConcreteBuf y))) = keccak' y /= x
+    checkKecc _ = True
+
+substituteEWord :: Map (Expr EWord) W256 -> Prop -> Prop
+substituteEWord valMap p = mapProp go p
+  where
+    go :: Expr a -> Expr a
+    go orig@(Var _) = Lit (valMap ! orig)
+    go orig@(TxValue) = Lit (valMap ! orig)
+    go a = a
+
+
+substituteBuf :: Map (Expr Buf) BufModel -> Prop -> Prop
+substituteBuf valMap p = mapProp go p
+  where
+    go :: Expr a -> Expr a
+    go orig@(AbstractBuf _) = case (valMap !? orig) of
+                                Just (Flat x) -> ConcreteBuf x
+                                Just (Comp _) -> internalError "No compressed allowed in fuzz"
+                                Nothing -> orig
+    go a = a
+
+substituteStores ::  Map (Expr 'EAddr) (Map W256 W256) -> Prop -> Prop
+substituteStores valMap p = mapProp go p
+  where
+    go :: Expr a -> Expr a
+    go (AbstractStore a) = case valMap !? a of
+                                  Just m -> ConcreteStore m
+                                  Nothing -> ConcreteStore mempty
+    go a = a
+
+-- Var extraction
+data CollectVars = CollectVars {vars :: Set.Set (Expr EWord)
+                               ,vals :: Set.Set W256
+                               }
+  deriving (Show)
+
+initVarsState :: CollectVars
+initVarsState = CollectVars {vars = Set.empty
+                            ,vals =  Set.empty
+                            }
+
+findVarProp :: Prop -> State CollectVars Prop
+findVarProp p = mapPropM go p
+  where
+    go :: forall a. Expr a -> State CollectVars (Expr a)
+    go = \case
+      e@(Var a) -> do
+        s :: CollectVars <- get
+        put CollectVars {vars=Set.insert (Var a) s.vars, vals = s.vals}
```

Comment:
Maybe use record update here:

```suggestion
        put (s { vars = Set.insert (Var a) s.vars })
```

Reply:
Yeah. I wanted to, but I can't. Try and it fails. No idea why. Haskell is mysterious. Maybe you can debug? In the meanwhile, I can't fix it, the error makes no sense, and what you wrote makes 100% sense but Haskell doesn't like it. I'd be happy if you could figure out why.... I really wanna know.

Reply:
It says:

```
src/EVM/Fuzz.hs:106:15: error:
    • Record update is ambiguous, and requires a type signature
    • In the second argument of ‘($)’, namely
        ‘s {vars = Set.insert (Var a) s.vars}’
      In a stmt of a 'do' block:
        put $ s {vars = Set.insert (Var a) s.vars}
      In the expression:
        do s :: CollectVars <- get
           put $ s {vars = Set.insert (Var a) s.vars}
           pure e
    |
106 |         put $ s {vars=Set.insert (Var a) s.vars}
```